File: Bot/CompleteTypeScript/src/Game.ts
import { GameResultError } from "./GameResultError";
import { DirectedAction, Result, State } from "./generated/swoq";
import { IGameServiceClient } from "./generated/swoq.client";

export class Game {
    private readonly client: IGameServiceClient;

    public readonly gameId: string;
    public readonly width: number;
    public readonly height: number;
    public readonly visibilityRange: number;
    public state: State;

    constructor(
        client: IGameServiceClient,
        gameId: string,
        width: number,
        height: number,
        visibilityRange: number,
        state: State
    ) {
        this.client = client;
        this.gameId = gameId;
        this.width = width;
        this.height = height;
        this.visibilityRange = visibilityRange;
        this.state = state;
    }

    public async act(action: DirectedAction): Promise<State> {
        const { response } = await this.client.act({
            gameId: this.gameId,
            action,
        });
        if (response.state !== undefined) {
            this.state = response.state;
        }
        if (response.result !== Result.OK) {
            // TODO what to do with this.state here? (Esp if/when it is undefined on response)
            throw new GameResultError(response.result);
        }
        return this.state;
    }
}

File: Bot/CompleteCSharp/Game.cs
using Swoq.Interface;

namespace Bot;

internal class Game(GameService.GameServiceClient client, StartResponse response, ReplayFile? replayFile) : IDisposable
{
    public void Dispose()
    {
        replayFile?.Dispose();
    }

    public string GameId { get; } = response.GameId;
    public int MapWidth { get; } = response.MapWidth;
    public int MapHeight { get; } = response.MapHeight;
    public int VisibilityRange { get; } = response.VisibilityRange;
    public State State { get; private set; } = response.State;
    public int? Seed { get; } = response.HasSeed ? response.Seed : null;

    public ActResult Act(DirectedAction action1, DirectedAction action2)
    {
        var request = new ActRequest() { GameId = GameId, Action = action1, Action2 = action2 };

        var response = client.Act(request);

        replayFile?.Append(request, response);

        //if (response.Result != ActResult.Ok)
        //{
        //    var status = response.State?.Status;
        //    throw new GameException($"Act failed (result {response.Result}, status {status})");
        //}

        if (response.State != null)
        {
            State = response.State;
        }

        return response.Result;
    }
}
